### PR TITLE
[transformer] Port Sequence Parallel

### PR DESCRIPTION
**Pull request by crcrpar**
- `tests/L0/run_transformer/test_transformer_module.py` is failing therefore is being skipped due to the lack of appropriate updates in apex/transformer/testing.
- apex.transformer expects tensors to be 3D and have the shape of (sequence, batch, feature).
- some operations in sequence parallel region takes care of certain chunk of tensors split in the dimension of sequence.
- gradients of params in sequence parallel region should be all_reduce'ed.

Ref: https://github.com/NVIDIA/Megatron-LM/releases/tag/v3.0.2

### CODE REVIEW

File: apex/transformer/testing/commons.py [hunk 1]
```diff
@@ -94,6 +202,36 @@ def loss_func(x):
...
+        model: torch.nn.Module,
+    ) -> Tuple[torch.Tensor, Callable[[torch.Tensor], Tuple[torch.Tensor, Dict[str, torch.Tensor]]]]:
+        x = batch[0] if isinstance(batch, list) else batch
+        if isinstance(x, torch.Tensor):
+            x = x.transpose(0, 1).contiguous()
+            if self.sequence_parallel_enabled:
+                x = x.chunk(
+                    chunks=transformer.parallel_state.get_tensor_model_parallel_world_size(),
+                    dim=0
+                )[transformer.parallel_state.get_tensor_model_parallel_rank()].contiguous()
```
Comment (crcrpar) [inline apex/transformer/testing/commons.py:222]: is this more concise?
```suggestion
                x = tensor_parallel.mappings.scatter_to_sequence_parallel_region(x)
```

File: tests/L0/run_transformer/test_layers.py [hunk 2]
```diff
@@ -171,139 +270,274 @@ def test_affine_weight_init_row_parallel_gpu(self) -> None:
...
+            self.BATCH_SIZE,
+            self.HIDDEN_SIZE,
+        )
+        for tensor_model_parallel_world_size in range(
+            1 + int(sequence_parallel_enabled), self.world_size + 1
+        ):
             if self.world_size % tensor_model_parallel_world_size:
                 continue
+            if sequence_parallel_enabled and tensor_model_parallel_world_size == 1:
+                continue
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_layers.py:304]: `test_row_parallel_linear_sequence_parallel` has a decorator which obviates this.
```suggestion
```

File: tests/L0/run_transformer/test_layers.py [hunk 3]
```diff
@@ -171,139 +270,274 @@ def test_affine_weight_init_row_parallel_gpu(self) -> None:
...
+        accumulation_in_fp16: bool,
+        sequence_parallel_enabled: bool,
     ):
         for tensor_model_parallel_world_size in range(1, self.world_size + 1):
-            with self.subTest(
-                tensor_model_parallel_world_size=tensor_model_parallel_world_size
-            ):
+            if async_tensor_model_parallel_allreduce and sequence_parallel_enabled:
+                if tensor_model_parallel_world_size == 1:
+                    continue
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_layers.py:436]: err, I can't remember the rationale for this condition.

File: tests/L0/run_transformer/test_transformer_module.py [hunk 4]
```diff
@@ -66,6 +66,8 @@ def run_transformer_tests():
             tensor_model_parallel_size = 1 + (1 - (num_devices % 2 and num_devices > 4))
             pipeline_model_parallel_size = num_devices // tensor_model_parallel_size
             test_run_cmd += f" --pipeline-model-parallel-size {pipeline_model_parallel_size} --tensor-model-parallel-size {tensor_model_parallel_size}"
+            # TODO(mkozuki): Update apex.transformer.testing
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_transformer_module.py:69]: verbosity?
```suggestion
            # TODO(mkozuki): Update apex.transformer.testing. The model def.s and global vars are outdated so 
            #     that gpt/bert script functionally fail. e.g. the order of tensor dims.
```

File: apex/transformer/pipeline_parallel/p2p_communication.py [hunk 5]
```diff
@@ -130,39 +144,69 @@ def _communicate(
...
 
         - tensor_recv_prev: `torch.Tensor` if `recv_prev` is :obj:`True`, `None` otherwise.
         - tensor_recv_next: `torch.Tensor` if `recv_next` is :obj:`True`, `None` otherwise.
     """
+    if async_comm and sequence_parallel_enabled:
+        import warnings  # NOQA
+        class ExperimentalWarning(UserWarning): pass  # NOQA
+        warnings.warn(
+            "The combination of `async_comm` and `sequence_parallel_enabled` is not well tested.",
+            ExperimentalWarnings,
```
Comment (crcrpar) [inline apex/transformer/pipeline_parallel/p2p_communication.py:162]: typo
```suggestion
            ExperimentalWarning,
```

**Comment by crcrpar** [comment]
9d0e65028100f889b75bd05d76ffd032e1dd9c97 for tests to be green even in an environment with torch_ucc.

**Comment by crcrpar** [comment]
1400 includes everything this has.
